### DISCUSSION TITLE
nodal variables do not have gradients

### DISCUSSION THREAD

**Discussion by gj19866** [idaholab/moose#26693] [Q&A General] [answered]
Dear MOOSErs,

I was hoping someone might be able to provide some assistance.
I would like to write an auxvariable that is calculated by the equation,
$\underline{j}=\psi_R \nabla \psi_I- \psi_I \nabla \psi_R -\nabla \varphi$, where $\psi_R$, $\psi_I$, and $\varphi$ are variables calculated in my programme. (N.B. $\underline{j}$ is a vector)

To do this I initially wrote an auxkernel using the ADVariables, and it failed to compile with the error:
```
error: cannot convert 'const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true>' to 'double' in assignment
```
I believed this was due to auxkernels not being able to handle ADvariables, so I rewrote the .C file with normal variables instead.

The rewritten file then compiled fine, however when running I get the error: 
`nodal variables do not have gradients`

This confused me slightly as the gradients for these variables are computed and used by some of the other kernels.

I know that it is possible for me to write a kernel to solve this equation, but as the other variables are all solved for already in the code this doesn't seem like a good solution. 

Please may you advise:
- If (and then how) you can use ADVectorAuxVariables
- If not how to go about solving this equation given that they are nodal variables, and these "do not have gradients"

I appreciate your patience.

Thank you very much,
Gillian

**Comment by GiudGiud** [comment, marked as answer]
Hello

AuxKernels do not support automatic differentiation. The dependency on nonlinear variables can be handled using material properties (or functor material properties if you are using FV, or complex FE schemes). 

Nodal variables do indeed not have a single defined value for the gradient at a node. At the node, there is a dicontinuity in the derivative, because for example one of the shape functions (the one non-zero at the node, at least for lagrange)'s support domain is centered on a node and its derivative changes sign at the node.

You can compute this quantity as an elemental quantity instead. For the auxvariable, use a variable that is defined on degrees of freedom that are on the element rather than on the node. For example, a constant monomial is defined by its elemental value (which we arbitrarily place at the centroid of the element in moose). 

You can use this as the auxvariable if the quantity should be a vector computed by a VectorAuxKernel
```
[AuxVariables]
[j]
  family = MONOMIAL_VEC
  order = .... (constant to start with?)
[]
```

Guillaume

**Reply by gj19866** [reply]
Amazing, thank you very much Guillaume!

Out of interest, is there a particular reason why AuxVars does not support AD? 
Is it just something that has not been implemented (or doesn't need to be!), or is there something preventing AuxVars being able to being AD?

All the best from the UK,
Gillian

**Reply by GiudGiud** [reply]
@lindsayad if you want to comment on why auxvariables do not support AD.

I would just say it s not needed as you can use material properties. It also does not seem efficient to me. When you compute an auxvariable at a Qp, it depends on the variable value at the dofs all around it. Which would themselves depend on nonlinear variable values around them (especially if the FE families to represent the nonlinear and auxiliary variables are different). So just like that you would get a huge AD stencil, in a very artificial way.

**Reply by lindsayad** [reply]
The problem with AD in the aux system is how to store the derivatives in the aux system solution vector. You cannot use a `PetscVector` as it only understands reals or complex numbers. Last time I thought about this I considered a `DistributedVector` but then stopped for some reason. As @GiudGiud said it has not been a priority since we have other means to compute "auxiliary" quantities with derivatives via the material system (and then these can also be auto-output into aux vars for postprocessing/visualization)

**Reply by gj19866** [reply]
That's really interesting. Thank you for taking the time to explain it to me.

**Reply by lindsayad** [reply]
No problem! 

I think I found the reason I stopped my last half-hearted attempt to do AD with aux, digging in my DM history with @roystgnr. The `System` class in libMesh holds `NumericVector<Number>` in its `_vectors` container, so even if `Number` was configurable to something other than `Real` or `std::complex<Real>` (which it currently can't be), you could not mix storage of vectors holding non-dual-numbers and dual-numbers. I think we could figure out a way to handle this, but it hasn't been worth the gains
